pkg/manager/impl/util/single_task_execution.go

```diff
@@ -0,0 +1,340 @@
+	bindings := make([]*core.Binding, 0, len(inputTemplate.Variables))
+	for key, val := range inputTemplate.Variables {
+		binding := &core.Binding{
+			Var: key,
+		}
+		var bindingData core.BindingData
+		if val.Type.GetSimple() != core.SimpleType_NONE {
+			if inputs.Literals[key] != nil {
+				bindingData = core.BindingData{
+					Value: &core.BindingData_Scalar{
+						Scalar: inputs.Literals[key].GetScalar(),
+					},
+				}
+			}
+
+		} else if val.Type.GetSchema() != nil {
+			if inputs.Literals[key] != nil && inputs.Literals[key].GetScalar() != nil {
+				bindingData = core.BindingData{
+					Value: &core.BindingData_Scalar{
+						Scalar: &core.Scalar{
+							Value: &core.Scalar_Schema{
+								Schema: inputs.Literals[key].GetScalar().GetSchema(),
+							},
+						},
+					},
+				}
+			}
+		} else if val.Type.GetCollectionType() != nil {
+			if inputs.Literals[key] != nil && inputs.Literals[key].GetCollection() != nil &&
+				inputs.Literals[key].GetCollection().GetLiterals() != nil {
+				collectionBindings := make([]*core.BindingData, len(inputs.Literals[key].GetCollection().GetLiterals()))
+				for idx, literal := range inputs.Literals[key].GetCollection().GetLiterals() {
+					collectionBindings[idx] = getBinding(literal)
+
+				}
+				bindingData = core.BindingData{
+					Value: &core.BindingData_Collection{
+						Collection: &core.BindingDataCollection{
+							Bindings: collectionBindings,
+						},
+					},
+				}
+			}
+		} else if val.Type.GetMapValueType() != nil {
+			if inputs.Literals[key] != nil && inputs.Literals[key].GetMap() != nil &&
+				inputs.Literals[key].GetMap().Literals != nil {
+				bindingDataMap := make(map[string]*core.BindingData)
+				for k, v := range inputs.Literals[key].GetMap().Literals {
+					bindingDataMap[k] = getBinding(v)
+				}
+
+				bindingData = core.BindingData{
+					Value: &core.BindingData_Map{
+						Map: &core.BindingDataMap{
+							Bindings: bindingDataMap,
+						},
+					},
+				}
+			}
+		} else if val.Type.GetBlob() != nil {
+			if inputs.Literals[key] != nil && inputs.Literals[key].GetScalar() != nil {
+				bindingData = core.BindingData{
+					Value: &core.BindingData_Scalar{
+						Scalar: &core.Scalar{
+							Value: &core.Scalar_Blob{
+								Blob: inputs.Literals[key].GetScalar().GetBlob(),
+							},
+						},
+					},
+				}
+			}
+		} else {
+			return nil, errors.NewFlyteAdminErrorf(codes.InvalidArgument, "Unrecognized value type [%+v]", val.GetType())
+		}
+		binding.Binding = &bindingData
+		bindings = append(bindings, binding)
+	}
+	return bindings, nil
+}
+
+func CreateOrGetWorkflowModel(
+	ctx context.Context, request admin.ExecutionCreateRequest, db repositories.RepositoryInterface,
+	workflowManager interfaces.WorkflowInterface, namedEntityManager interfaces.NamedEntityInterface, taskIdentifier *core.Identifier,
+	task *admin.Task) (*models.Workflow, error) {
+	workflowModel, err := db.WorkflowRepo().Get(ctx, repositoryInterfaces.GetResourceInput{
```

Comment:
Doesn't this introduce a race condition? if I tried to run the same task twice, both of them go past the "Get" call and realize there isn't an existing WF model, so they will both try to create one... Should the error check for Create() call below exclude "ErrExistingRow" or whatever the error is for that?

Reply:
yes, good call. done